File: src/test/java/DemoQATest.java
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.interactions.WheelInput;
import org.openqa.selenium.support.ui.Wait;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

import static org.junit.Assert.assertEquals;

public class DemoQATest {
    private void scrollPastElement(WebDriver driver, WebElement element) {
	WheelInput.ScrollOrigin scrollOrigin = WheelInput.ScrollOrigin.fromElement(element);
	new Actions(driver).scrollFromOrigin(scrollOrigin, 0, 200).perform();
    }

    @Test
    public void test() {
        final String USERNAME = "research_paper";
        final String PASSWORD = "1tq52rw43E!";

	ChromeOptions options = new ChromeOptions();
	options.addArguments("--headless=new");
        WebDriver driver = new ChromeDriver(options);
        driver.manage().timeouts().implicitlyWait(Duration.ofMillis(4000));
        Wait<WebDriver> wait = new WebDriverWait(driver, Duration.ofSeconds(10));

        // Load the login page
        driver.get("https://demoqa.com/login");

        // Input username and password, click login button
        driver.findElement(By.id("userName")).sendKeys(USERNAME);
        driver.findElement(By.id("password")).sendKeys(PASSWORD);
        driver.findElement(By.id("login")).click();

        // Click "Go to Store" button
	WebElement storeButton = driver.findElement(By.id("gotoStore"));
	scrollPastElement(driver, storeButton);
        storeButton.click();

        // Click on the first book title
        driver.findElement(By.id("see-book-Git Pocket Guide")).click();

        // Store the name of the book
        String bookName = driver.findElement(By.cssSelector("#title-wrapper #userName-value")).getText();

        // Add the book to the collection
	WebElement addNewRecordButton = driver.findElement(By.id("addNewRecordButton"));
        addNewRecordButton.click();

        // Click profile button
	WebElement profileButton = driver.findElement(By.cssSelector(".show #item-3"));
	scrollPastElement(driver, profileButton);
        profileButton.click();

        // Wait until the profile page is loaded. This takes a while.
        WebElement profileElement = driver.findElement(By.id("books-wrapper"));
        wait.until(d -> profileElement.isDisplayed());

        // Click on the first book in the profile table
        driver.findElement(By.id("see-book-Git Pocket Guide")).click();

        // Get the book title and compare it to the stored value
        String firstRowTitleText = driver.findElement(By.cssSelector("#title-wrapper #userName-value")).getText();
        assertEquals(firstRowTitleText, bookName);

        // Go back to the profile page and delete the book.
        driver.navigate().back();
        driver.findElement(By.id("delete-record-undefined")).click();

        driver.close();
    }
}

File: src/test/python/DemoQATest.py
import unittest
from selenium import webdriver
from selenium.webdriver import ActionChains
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


class DemoQATest(unittest.TestCase):
    def scroll_past_element(self, driver, element):
        _ = element.location_once_scrolled_into_view
        ActionChains(driver).move_to_element(element).perform()

    def test_demoqa(self):
        USERNAME = "research_paper"
        PASSWORD = "1tq52rw43E!"

        dimension = (1920, 1080)
        options = Options()
        options.add_argument("--headless=new")
        driver = webdriver.Chrome(options=options)
        driver.implicitly_wait(4)  # seconds
        driver.set_window_size(*dimension)
        wait = WebDriverWait(driver, 10)  # seconds

        # Load the login page
        driver.get("https://demoqa.com/login")

        # Input username and password, click login button
        driver.find_element(By.ID, "userName").send_keys(USERNAME)
        driver.find_element(By.ID, "password").send_keys(PASSWORD)
        driver.find_element(By.ID, "login").click()

        # Click "Go to Store" button
        store_button = driver.find_element(By.ID, "gotoStore")
        self.scroll_past_element(driver, store_button)
        store_button.click()

        # Click on the first book title
        driver.find_element(By.ID, "see-book-Git Pocket Guide").click()

        # Store the name of the book
        book_name = driver.find_element(By.CSS_SELECTOR, "#title-wrapper #userName-value").text

        # Add the book to the collection
        add_new_record_button = driver.find_element(By.ID, "addNewRecordButton")
        add_new_record_button.click()

        # Click profile button
        profile_button = driver.find_element(By.CSS_SELECTOR, ".show #item-3")
        self.scroll_past_element(driver, profile_button)
        profile_button.click()

        # Wait until the profile page is loaded. This takes a while.
        profile_element = driver.find_element(By.ID, "books-wrapper")
        wait.until(EC.visibility_of(profile_element))

        # Click on the first book in the profile table
        driver.find_element(By.ID, "see-book-Git Pocket Guide").click()

        # Get the book title and compare it to the stored value
        first_row_title_text = driver.find_element(By.CSS_SELECTOR, "#title-wrapper #userName-value").text
        self.assertEqual(first_row_title_text, book_name)

        # Go back to the profile page and delete the book.
        driver.back()
        driver.find_element(By.ID, "delete-record-undefined").click()

        driver.close()


if __name__ == "__main__":
    unittest.main()
